docs/internal/expressions.md

```diff
@@ -0,0 +1,611 @@
+  b = ...; // ERROR
+
+  // Updates a, still valid
+  a = 38;
+} // `a` goes out of scope
+```
+
+### Object fields
+
+Like other variables, object fields store pointers to objects, unless declared `embed`, which makes them part of the structure.
+Refer to the types document for more information on the `embed` keyword.
+
+## Operators
+
+In Verona, operators and calls are treated equally.
+In-fix operators (ex. `a + b`) and pre-fix operators (ex. `+(a, b)`) are one and the same.
+
+All operators are calls on a descriptor: either an object or a type declaration that indicates which type the method is to be found.
+For example the type name (`Foo::method()`), the first argument of a pre-fix operator (`foo.method()` or `method(foo)`), or the left hand side of an infix operator (`foo method bar`).
+
+Example:
+```ts
+// Defines the type of `x`
+let x : U32 & imm = 42;
+
+// Infers the type of `y` and `10` from the type of `x`
+// Checks that `U32::+(U32, U32)` exists
+// Calls U32::+(x, 10)
+let y = x + 10;
+
+// The above is identical to ones below
+let a = x.+(10);
+let b = U32::+(x, 10);
+```
+
+The Verona standard library will implement all operations for all appropriate types, including all arithmetic for all numeric types and string handling for all character and string types.
+Users can implement similar calls to specialised type (ex. `MyInt::+(MyInt, I32)`).
+
+There is no precedence in Verona, so operators must be wrapped in parenthesis to convey the right meaning.
+This means that, unlike other languages, `(a + b * c)` is a syntax error, while either `((a + b) * c)` or `(a + (b * c))` are expected.
+
+_Note: Can we avoid forcing parenthesis on DSLs?_
+
+## Arithmetic
+
+As detailed above, arithmetic in Verona is implemented as function calls on numeric types and operands.
+This means developers can extend the functionality of arithmetic to the existing types or their own new types naturally.
+But it also means that every addition or subtraction could end up as call on objects, which could potentially be many orders of magnitude slower than standard hardware operations.
+
+But Verona's arithmetic is often as fast as calling the right hardware instructions.
+
+The reasons are:
+1. As explained in the types document, numeric types are singleton classes, ie. they don't have fields, only methods.
+   The compiler treats them differently and uses their machine representation instead of a pointer to an allocated memory.
+2. Computation often uses concrete types (ex. `U32` or `F64`) instead of interface types (ex. `IntegralType`) or type unions (ex. `(U32 | F64`)).
+   The compiler can only match concrete types to their machine equivalent.
+   It can try to infer, match and separate concrete types from their declared collection types, but that's an optimisation, not a guarantee.
+3. Direct code generation of the hardware instructions for arithmetic functions.
+   Those functions are often short and simple and are always inlined by the compiler.
```

Comment:
```suggestion
3. The functions that implement arithmetic operations are trivial (often single instruction) functions that are ideal inlining candidates.

The combination of these means that a `U32` can be stored as a 32-bit integer, rather than a pointer to a 32-bit integer stored elsewhere, that the `+` operation can be a direct call (no late binding or dynamic dispatch) to a built-in function that implements 32-bit integer addition, and this function can always be inlined.
This comes without any loss of generality: you can still use dynamic dispatch over numeric types via interfaces or union types but you only pay for the additional overhead when you use it.
```